src/object/cli_obj.c

```diff
@@ -2340,26 +2347,83 @@ obj_null_csum(const daos_obj_update_t *args) {
 	}
 }
 
-static void
-obj_update_csums(const struct dc_object *obj, const daos_obj_update_t *args) {
+static int
+csum_obj_update(struct dc_object *obj, daos_obj_update_t *args)
+{
 	struct daos_csummer	*csummer = dc_cont_hdl2csummer(obj->cob_coh);
-	daos_iod_t		*iod;
 	int			 i;
+	int			 rc;
 
 	obj_null_csum(args);
 	if (!daos_csummer_initialized(csummer)) /** Not configured */
-		return;
+		return 0;
+
+	/** Calc 'd' key checksum */
+	rc = daos_csummer_calc_key(csummer, args->dkey, &args->dkey_csum);
+	if (rc != 0)
+		return rc;
 
+	if (DAOS_FAIL_CHECK(DAOS_CHECKSUM_UPDATE_DKEY_FAIL))
+		((char *) args->dkey->iov_buf)[0]++;
+
+	/** Calc 'a' key checksum and value checksum */
 	for (i = 0; i < args->nr; i++) {
-		iod = &args->iods[i];
+		daos_iod_t *iod = &args->iods[i];
 
 		if (!csum_iod_is_supported(csummer->dcs_chunk_size, iod))
 			continue;
-		daos_csummer_calc(csummer, &args->sgls[i],
+		rc = daos_csummer_calc(csummer, &args->sgls[i],
 				  iod, &iod->iod_csums);
+		if (rc != 0)
+			return rc;
+
 		if (DAOS_FAIL_CHECK(DAOS_CHECKSUM_UPDATE_FAIL))
 			((char *) iod->iod_csums->cs_csum)[0]++;
+
+		rc = daos_csummer_calc_key(csummer, &iod->iod_name,
+					   &iod->iod_kcsum);
+		if (rc != 0)
```

Comment:
When the buffer that is allocated in above steps can be freed for the failure case?